Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -262,6 +262,55 @@ class BrowserViewController: UIViewController {
                 self.tabManager.addTabAndSelect(request, isPrivate: PrivateBrowsingManager.shared.isPrivateBrowsing)
             }
         }
+        
+        backgroundDataSource.initializeFavorites = { sites in
+            DispatchQueue.main.async {
+                defer { Preferences.NewTabPage.preloadedFavoritiesInitialized.value = true }
+                
+                if Preferences.NewTabPage.preloadedFavoritiesInitialized.value
+                    || Bookmark.hasFavorites { return }
+                
+                guard let sites = sites, sites.count > 0 else {
+                    FavoritesHelper.addDefaultFavorites()
+                    return
+                }
+                
+                let customFavorites = sites.compactMap { $0.asFavoriteSite }
+                Bookmark.addFavorites(from: customFavorites)
+            }
+        }
+        
+        backgroundDataSource.replaceFavoritesIfNeeded = { sites in
+            if Preferences.NewTabPage.initialFavoritesHaveBeenReplaced.value { return }
+            
+            guard let sites = sites, sites.count > 0 else { return }
+            
+            DispatchQueue.main.async {
+                let defaultFavorites = PreloadedFavorites.getList()
+                let currentFavorites = Bookmark.allFavorites
+                
+                if defaultFavorites.count != currentFavorites.count {
+                    return
+                }
+                
+                let exactSameFavorites = Bookmark.allFavorites .filter {
+                    guard let urlString = $0.url, let url = URL(string: urlString),
```

Comment:
If adding a force linebreak for the 3rd item, should probably just put them all on their own line.